File: provision.py
#!/usr/bin/env python3
import os
import subprocess
from abc import abstractmethod, ABCMeta
from typing import List


def main():
    Provisioner([

        # brew
        InstallHomeBrewTask(),
        BrewTask('peco'),
        BrewTask('fzf'),
        BrewTask('mas'),

        # brew cask
        InstallHomeBrewCaskTask(),
        BrewCaskTask('slack'),
        BrewCaskTask('karabiner-elements'),
        BrewCaskTask('alfred'),
        BrewCaskTask('iterm2'),
        BrewCaskTask('caffeine'),
        BrewCaskTask('google-japanese-ime'),

        # mas
        MasTask(409183694),  # keynote
        MasTask(409203825),  # Numbers
        MasTask(409201541),  # Pages
        MasTask(539883307),  # LINE
        MasTask(409183694),  # Keynote
        MasTask(485812721),  # TweetDeck
        MasTask(405399194),  # Kindle

        # vim

    ]).provision()


class Provisioner:
    def __init__(self, tasks: List['Task']):
        self.tasks = tasks

    def provision(self):
        for task in self.tasks:
            task.provision_if_not()


class Task(metaclass=ABCMeta):
    @abstractmethod
    def is_provisioned(self) -> bool:
        pass

    @abstractmethod
    def provision(self) -> None:
        pass

    def to_string(self) -> str:
        return str(self.__class__.__name__) + str(self.__dict__)

    def provision_if_not(self, level: int = 0) -> None:
        print(("    " * level) + bold("TASK"), self.to_string(), "->", end=' ')

        if self.is_provisioned():
            print(green("already provisioned"))
            return

        print(red("not provisioned"))
        self.provision()


red = lambda text: '\033[1;31m' + text + '\033[0m'
green = lambda text: '\033[1;32m' + text + '\033[0m'
grey = lambda text: '\033[0;32m' + text + '\033[0m'
bold = lambda text: '\033[1;34m' + text + '\033[0m'


class InstallHomeBrewTask(Task):

    def is_provisioned(self) -> bool:
        return is_command_exists('brew')

    def provision(self) -> None:
        cmd_exec(['/usr/bin/ruby', '-e',
                  '"$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"'])


class InstallHomeBrewCaskTask(Task):

    def is_provisioned(self) -> bool:
        return not cmd_has_error(['brew', 'cask'])

    def provision(self) -> None:
        cmd_exec(['brew', 'tap', 'caskroom/cask'])


class MasTask(Task):

    def __init__(self, app_id: int):
        self.app_id: int = app_id

    def is_provisioned(self) -> bool:
        return not cmd_has_error(["mas list | grep '^{} '".format(self.app_id)])

    def provision(self) -> None:
        cmd_exec(['mas', 'install', str(self.app_id)])


class BrewTask(Task):
    PACKAGE_BASE_DIR = "/usr/local/Cellar"

    def __init__(self, package: str, install_options: List[str] = None):
        self.package: str = package
        self.install_options: List[str] = install_options or []

    def is_provisioned(self) -> bool:
        # FIXME difference of install_option is not detected
        package_path = os.path.join(__class__.PACKAGE_BASE_DIR, self.package)
        return not cmd_has_error(['ls', package_path])

    def provision(self) -> None:
        cmd_exec(['brew', 'install', self.package] + self.install_options)


class BrewCaskTask(Task):
    PACKAGE_BASE_DIR = "/usr/local/Caskroom"

    def __init__(self, package: str, install_options: List[str] = None):
        self.package: str = package
        self.install_options: List[str] = install_options or []

    def is_provisioned(self) -> bool:
        # FIXME difference of install_option is not detected
        package_path = os.path.join(__class__.PACKAGE_BASE_DIR, self.package)
        return not cmd_has_error(['ls', package_path])

    def provision(self) -> None:
        cmd_exec(['brew', 'install', self.package] + self.install_options)


def is_command_exists(command: str) -> bool:
    return not cmd_has_error(['which', command])


def cmd_has_error(args: List[str]) -> bool:
    result = subprocess.run(args=args, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
    return result.returncode != 0


def cmd_exec(args: List[str]) -> subprocess.CompletedProcess:
    return subprocess.run(args=args, check=True, shell=True)


if __name__ == '__main__':
    main()

File: provision.rb
#!/usr/bin/env ruby

# todo debug mode
# todo dry-run mode
# todo cannot interrupt brew cask install
require './provision-task'

tasks do
  task :foge do
    task :fuga do
      task :foo, "echo 'foo'"
      task :bar, "echo 'bar'"
    end
  end
end

raise "a"

tasks do

  task :default do
    task :init do
      task :install_homebrew, if_err('which brew'),
           '/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"'
      task :tap_brew_cask, if_err('brew tap | grep caskroom/cask'), 'brew tap caskroom/cask'

      task brew 'git'
      task symlink '~/.dotfiles/git/gitignore', '~/.gitignore'

      ghq_root = '~/repos'

      task :dotfiles_repo do
        host = 'github.com'
        repo = 'ikenox/dotfiles'
        dotfiles_origin_dir = "#{ghq_root}/#{host}/#{repo}"
        task :clone_dotfiles, if_not_exist(dotfiles_origin_dir), "git clone git@#{host}:#{repo}.git #{dotfiles_origin_dir}"
        task symlink dotfiles_origin_dir, '~/.dotfiles'
      end
      task :gitconfig do
        task symlink '~/.dotfiles/git/gitconfig', '~/.gitconfig'
        task "git config -f ~/.gitconfig.local ghq.root #{ghq_root}"
      end
    end

    task :setup_git do
      task :set_username, if_err("git config user.name"), -> {
        print "please type your git user.name: "
        name = gets.chomp
        "git config -f ~/.gitconfig.local user.name '#{name}'"
      }
      task :set_email, if_err("git config user.email"), -> {
        print "please type your git user.email: "
        email = gets.chomp
        "git config -f ~/.gitconfig.local user.email '#{email}'"
      }
      task brew 'ghq'
    end

    task :karabiner_elements do
      task brew_cask 'karabiner-elements'
      task symlink '~/.dotfiles/karabiner', '~/.config/karabiner'
    end

    task :setup_vim do
      task brew 'vim'
      task :install_vim_plug, if_not_exist("~/.vim/autoload/plug.vim"),
           "curl -fLo ~/.vim/autoload/plug.vim --create-dirs https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim"
      task symlink '~/.dotfiles/vim/vimrc', '~/.vimrc'
      task symlink '~/.dotfiles/vim/vimrc.keymap', '~/.vimrc.keymap'
    end

    task brew 'peco'
    task brew 'fzf'
    task brew 'jq'
    task brew 'mas'
    task :ag do
      task brew 'the_silver_searcher'
      task symlink '~/.dotfiles/ag/agignore', '~/.agignore'
    end

    task :fish do
      task brew 'fish'
      task symlink '~/.dotfiles/fish', '~/.config/fish'
      task :fisherman, if_not_exist('~/.config/fish/functions/fisher.fish'),
        'curl -Lo ~/.config/fish/functions/fisher.fish --create-dirs git.io/fisher'
      task :set_default_shell do
        task :add_shell, if_err("cat /etc/shells | grep $(which fish)"),
          "sudo bash -c 'echo $(which fish) >> /etc/shells'"
        # task :add_shell, if_err("echo $SHELL | grep fish") do
        #   sh "sudo chsh -s $(which fish)"
        # end
      end
      task 'fish -c "fisher"'
    end

    task :hyper do
      task brew_cask 'hyper'
      task symlink '~/.dotfiles/hyper/hyper.js', '~/.hyper.js'
    end

    # todo: set keyboard -> 入力ソース -> ひらがな(google)
    task brew_cask 'google-japanese-ime'

    task :osx_defaults do
      task 'defaults write com.apple.dock autohide -bool true'
      task 'defaults write com.apple.dock persistent-apps -array'
      task 'defaults write com.apple.dock tilesize -int 55'
      task 'defaults write com.apple.dock wvous-bl-corner -int 10'
      task 'defaults write com.apple.dock wvous-bl-modifier -int 0'
      # todo killall if updated
      # killall Dock'

      task 'defaults write com.apple.finder AppleShowAllFiles YES'
      task 'defaults write com.apple.finder NewWindowTarget -string "PfDe"'
      task 'defaults write com.apple.finder NewWindowTargetPath -string "file://${HOME}/"'
      # todo killall if updated
      # killall Finder'

      task 'defaults write com.apple.Safari IncludeInternalDebugMenu -bool true'

      # todo needs restart
      task 'defaults write -g com.apple.trackpad.scaling -int 3'
      task 'defaults write -g InitialKeyRepeat -int 15'
      task 'defaults write -g KeyRepeat -int 2'
      task 'defaults -currentHost write -globalDomain com.apple.mouse.tapBehavior -int 1'
    end

    task :container_tools do
      task brew_cask 'virtualbox'
      task brew_cask 'vagrant'
      task if_err('vagrant plugin list | grep vagrant-vbguest'), 'vagrant plugin install vagrant-vbguest'
      task brew 'docker'
      task brew 'docker-machine'
      task brew 'docker-compose'
    end

    task :intellij do
      task brew_cask "jetbrains-toolbox"
      task symlink '~/.dotfiles/intellij/ideavimrc', '~/.ideavimrc'
      # TODO apply settings.jar
    end

    task :jupyter do
      task symlink '~/.dotfiles/jupyter/custom.js', '~/.jupyter/custom/custom.js'
    end

    task :python do
      task symlink '~/.dotfiles/matplotlib/matplotlibrc', '~/.matplotlibrc'
      task :pyenv do
        task brew 'pyenv'
        task brew 'pyenv-virtualenv'
        task symlink '~/.dotfiles/zsh/zshrc.module.pyenv', '~/.zshrc.module.pyenv'
      end
    end

    task :java do
      task brew 'jenv'
      task brew_cask 'java8'
      task 'echo "set PATH $HOME/.jenv/bin $PATH" > ~/.dotfiles/fish/conf.d/jenv.fish'
      task 'curl https://raw.githubusercontent.com/gcuisinier/jenv/master/fish/jenv.fish > ~/.config/fish/jenv.fish'
      task 'curl https://raw.githubusercontent.com/gcuisinier/jenv/master/fish/export.fish > ~/.config/fish/export.fish'
    end

    #task_brew_cask 'slack'
    task brew_cask 'alfred' # todo change hotkey from gui
    task brew_cask 'caffeine'
    task brew_cask 'discord'

    # tood need login app store
    task mas 409183694 # keynote
    task mas 409203825 # Numbers
    task mas 409201541 # Pages
    task mas 539883307 # LINE
    task mas 485812721 # TweetDeck
    task mas 405399194 # Kindle

    # todo macos
    # disable spotlight
    # かな入力
    # enable key repeat
    #
    # todo python
    # https://qiita.com/zreactor/items/c3fd04417e0d61af0afe
    # sudo installer -pkg /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg -target /
  end

  task :test do
    task :init do
      task :install_homebrew, if_err('which brew'), 'brew install hoge'
      task :install_homebrew, if_err('which brew'),
           '/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"'
      task :install_homebrew, if_err('which brew'),
           '/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"'
      task do
        task brew 'git'
      end

      task 'rm -rf foo/bar'
      task :hoge, 'rm -rf foo/bar'
      task :tap_brew_cask, if_not_exist('~/.foo'), 'brew tap caskroom/cask'

      task brew 'git'
      task brew 'git'

      task symlink '~/.dotfiles/git/gitignore', '~/.gitignore'
      task symlink '~/.dotfiles/git/gitignore', '~/.gitignore'
    end
  end
end
